Lima/darwin: Move docker socket to ~/.rd/docker.sock

mook-as opened the pull request:
Fixes #1841.
Conflicts with #1923.

jandubois commented on src/utils/__tests__/paths.spec.ts:
```diff
@@ -25,31 +25,43 @@ describe('paths', () => {
     },
+    altAppHome: {
+      win32:  '%APPDATA%/rancher-desktop/',
+      linux:  '%HOME%/.config/rancher-desktop/',
```
Shouldn't the Linux location also be `%HOME%/.rd/`?

jandubois replied:
It is just for consistency. When you talk to people on Slack, it is not always clear which platform they are on. So having the same answer for Linux and macOS is helpful. Also for docs, if it makes it into docs.

The same reason I want to be able to say `rdctl shell cat /etc/os-release` and have it work everywhere.

jandubois commented on src/k8s-engine/lima.ts:
```diff
@@ -1763,14 +1763,25 @@ export default class LimaBackend extends events.EventEmitter implements K8s.Kube
+      paths.logs,
+    ]);
+
+    if (!Array.from(pathsToDelete).some(dir => paths.lima.startsWith(dir))) {
```
Why do we need this? Is this triggered when someone moves the lima directory elsewhere and replaces it with a symlink?

I'm kind of conflicted about deleting a directory the user has created themselves. It will not interfere with the factory reset, because the symlink got deleted, and the new lima directory will be created in the default place again. So maybe it is better to just display a warning dialog in this case? Or ask for permission to delete this directory as well? Obviously in a different PR, but what do you think?

mook-as replied:
No; `paths.lima` is hard coded, it's just a filter because on mac, `paths.appHome` is a prefix of `paths.lima`.  However, this is _not_ the case on Linux, so we need to remove both.

jandubois replied:
I guess this could use some comment then? 😄

jandubois commented on src/k8s-engine/lima.ts:
```diff
@@ -1763,14 +1763,25 @@ export default class LimaBackend extends events.EventEmitter implements K8s.Kube
+      paths.altAppHome,
+      paths.config,
+      paths.logs,
+    ]);
```
I know it doesn't belong into this PR, but the list is also missing the cache directory for the `rancher-desktop-updater` (see #1684), and I'm wondering where/how that directory name is being constructed.

mook-as replied:
Ugh, right, but at least it's clearly a separate PR :D

(It's based on some Electron-internal logic, I think)

jandubois reviewed (approved):
LGTM

jandubois commented:
Testing shows that the socket will not be created when the directory doesn't exist:

```console
$ ls -l /var/run/docker.sock
lrwxr-xr-x  1 root  daemon  26 31 Mar 18:01 /var/run/docker.sock -> /Users/jan/.rd/docker.sock
$ ls -ld ~/.rd
ls: /Users/jan/.rd: No such file or directory
```

This should not be a problem once we create `~/.rd/bin` for the supporting utilities, but just so that we don't have to think about possible race conditions, the `docker.sock` code should make sure the directory exists before starting the VM.

jandubois commented on src/k8s-engine/lima.ts:
```diff
@@ -653,7 +653,7 @@ export default class LimaBackend extends events.EventEmitter implements K8s.Kube
     if (!dockerPortForwards) {
       config.portForwards?.push({
         guestSocket: '/var/run/docker.sock',
-        hostSocket:  'docker',
+        hostSocket:  path.join(paths.altAppHome, 'docker.sock'),
       });
     }
   }
```
This code only adds a portforwarding rule for `/var/run/docker.sock` if no such rule exists yet. But it will not update the rule when it already exists.

We should first delete any rule that forwards `/var/run/docker.sock` → `docker` first, so we'll keep any modified rule, but replace the old default rule.

BTW, I'm not quite sure why we need to keep any non-default rule; I think that should not be supported, and we should unconditionally force the socket to be forwarded to `~/.rd/lima`. If the user needs a socket elsewhere, they can always add a symlink. But let me know if I missed the reason we do this!

jandubois commented:
> Testing shows that the socket will not be created when the directory doesn't exist:

No, it doesn't show this. The reason nothing was created was because this PR does not update any existing forwarding rule for the docker socket; it only changes the host location when the rule is created for the first time.

mook-as commented:
Rebased; also, fixed the issue with existing VMs (created a VM on `main`, made sure the socket in `${LIMA_HOME}` exists, ran the new code, checked that the socket moved to `~/rd`).

ericpromislow reviewed (approved):
LGTM - I get the reasons for the changes, and it works fine. Was able to use `docker.sock` in the new location.

jandubois reviewed (approved):
Thanks, LGTM
